crates/red_knot_test/src/assertion.rs

```diff
@@ -0,0 +1,621 @@
+    inner: std::iter::Peekable<AssertionWithRangeIterator<'a>>,
+}
+
+impl<'a> Iterator for LineAssertionsIterator<'a> {
+    type Item = LineAssertions<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let file = self.file_assertions;
+        let ranged_assertion = self.inner.next()?;
+        let mut collector = AssertionVec::new();
+        let mut line = file.line_number(&ranged_assertion);
+        // Collect all own-line comments on consecutive lines; these all apply to the same line of
+        // code. For example:
+        //
+        //     # Error: [unbound-name]
+        //     # Type: Unbound
+        //     reveal_type(x)
+        //
+        if file.is_own_line(&ranged_assertion) {
+            collector.push(ranged_assertion.into());
+            let mut only_own_line = true;
+            while let Some(ranged_assertion) = self.inner.peek() {
+                let next_line = line.saturating_add(1);
+                if file.line_number(ranged_assertion) == next_line {
+                    if !file.is_own_line(ranged_assertion) {
+                        only_own_line = false;
+                    }
+                    line = next_line;
+                    collector.push(self.inner.next().unwrap().into());
+                    // If we see an end-of-line comment, it has to be the end of the stack,
+                    // otherwise we'd botch this case, attributing all three errors to the `bar`
+                    // line:
+                    //
+                    //     # Error:
+                    //     foo  # Error:
+                    //     bar  # Error:
+                    if !only_own_line {
+                        break;
+                    }
+                } else {
+                    break;
+                }
+            }
+            if only_own_line {
+                // The collected comments apply to the _next_ line in the code.
+                line = line.saturating_add(1);
+            }
+        } else {
+            // We have a line-trailing comment; it applies to its own line, and is not grouped.
+            collector.push(ranged_assertion.into());
+        }
+        Some(LineAssertions {
+            line,
+            assertions: collector,
+        })
+    }
+}
+
+impl std::iter::FusedIterator for LineAssertionsIterator<'_> {}
+
+/// One or more assertions referring to the same line of code.
+#[derive(Debug)]
+pub(crate) struct LineAssertions<'a> {
+    /// The line these assertions refer to.
+    ///
+    /// Not necessarily the same line the assertion comment is located on; for an own-line comment,
+    /// it's the next non-assertion line.
+    pub(crate) line: OneIndexed,
+
+    /// The assertions referring to this line.
+    pub(crate) assertions: AssertionVec<'a>,
+}
+
+impl<'a, 'b> From<&'a LineAssertions<'b>> for &'a [Assertion<'b>] {
+    fn from(value: &'a LineAssertions<'b>) -> Self {
+        &value.assertions[..]
+    }
+}
+
+static TYPE_RE: Lazy<Regex> =
```

Comment:
Yeah, in this case I chose to prioritize ease of implementation over performance :)

I looked at RegexSet but discarded it as an option because it doesn't support capture groups.